### DISCUSSION TITLE
How to handle a playlist file that references other files?

### DISCUSSION THREAD

**Discussion by Be-ing** [flatpak/xdg-desktop-portal#1288] [Questions]
I'm looking into https://github.com/flathub/org.mixxx.Mixxx/issues/45 regarding importing files referenced from a playlist file into Mixxx's database. As far as I can tell, there is no nice way to handle this use case with the existing file picker portal. Selecting the playlist file (via QFileDialog) works, and the sandboxed application can read the contents of the playlist file, but things get tricky when trying to open the file paths referenced in the playlist file.

There are two distinct cases: the playlist contains relative or absolute paths. In the case of absolute paths, conceivably the application could use the file picker portal to ask for permission for every individual file. This would be annoying and a subpar user experience, especially considering the application just opened the file picker to pick the playlist file immediately before this.

In the case of relative file paths, I don't know if there's any way this can be done currently. The application gets a remapped file path from the portal like `/run/user/1000/doc/8b2a07f4/playlist.m3u8`, so then the application would look for the referenced files in `/run/user/1000/doc/8b2a07f4/`, but there are no such files at that path because they haven't been remapped by the portal, nor is there a way to construct their path from the playlist's original, unremapped file path.

**Comment by Mikenux** [comment]
The proposal you commented on (https://github.com/flatpak/xdg-desktop-portal/discussions/1259#discussioncomment-8552270) is not the correct one because it involves getting files with the exact same name as the first file you open.

There are the following proposals, but they are limited to the folder you opened a file from:
- Simple Method. This is to provide access to all the contents of the folder. See https://github.com/flatpak/xdg-desktop-portal/issues/463#issuecomment-1867829867 for mockup.
- Application-defined Restrictions. See https://github.com/flatpak/xdg-desktop-portal/issues/463#issuecomment-1409305061.
- Maximum Sandbox. See https://github.com/flatpak/xdg-desktop-portal/discussions/1256. But I'll probably update this one to have better sandboxing (I'm currently updating another one).

However, as you said, playlist files can have absolute paths, but more importantly can require access to files from different locations. In this case there is no suitable proposal. Mine could possibly include this case, but I need to think about it more. Also note that I am not a developer and each of my proposals requires people to develop them.

**Reply by Be-ing** [reply]
> However, as you said, playlist files can have absolute paths, but more importantly can require access to files from different locations. In this case there is no suitable proposal.

Yes, a playlist file *could* do that, however I think it's unlikely in practice.
